coderd/notifications/reports/generator.go

```diff
@@ -0,0 +1,296 @@
+			return failedBuilds[i].TemplateVersionName < failedBuilds[j].TemplateVersionName
+		}
+		return failedBuilds[i].WorkspaceBuildNumber > failedBuilds[j].WorkspaceBuildNumber
+	})
+
+	// Build notification model for template versions and failed workspace builds
+	templateVersions := []map[string]any{}
+	for _, failedBuild := range failedBuilds {
+		c := len(templateVersions)
+
+		if len(templateVersions) == 0 || templateVersions[c-1]["template_version_name"] != failedBuild.TemplateVersionName {
+			templateVersions = append(templateVersions, map[string]any{
+				"template_version_name": failedBuild.TemplateVersionName,
+				"failed_count":          1,
+				"failed_builds": []map[string]any{
+					{
+						"workspace_owner_username": failedBuild.WorkspaceOwnerUsername,
+						"workspace_name":           failedBuild.WorkspaceName,
+						"build_number":             failedBuild.WorkspaceBuildNumber,
+					},
+				},
+			})
+			continue
+		}
+
+		//nolint:errorlint,forcetypeassert // only this function prepares the notification model
+		builds := templateVersions[c-1]["failed_builds"].([]map[string]any)
+		builds = append(builds, map[string]any{
+			"workspace_owner_username": failedBuild.WorkspaceOwnerUsername,
+			"workspace_name":           failedBuild.WorkspaceName,
+			"build_number":             failedBuild.WorkspaceBuildNumber,
+		})
+		templateVersions[c-1]["failed_builds"] = builds
+		//nolint:errorlint,forcetypeassert // only this function prepares the notification model
+		templateVersions[c-1]["failed_count"] = templateVersions[c-1]["failed_count"].(int) + 1
+	}
+
+	return map[string]any{
+		"failed_builds":     stats.FailedBuilds,
+		"total_builds":      stats.TotalBuilds,
+		"report_frequency":  frequencyLabel,
+		"template_versions": templateVersions,
+	}
+}
+
+func findTemplateAdmins(ctx context.Context, db database.Store, stats database.GetWorkspaceBuildStatsByTemplatesRow) ([]database.GetUsersRow, error) {
+	users, err := db.GetUsers(ctx, database.GetUsersParams{
+		RbacRole: []string{codersdk.RoleTemplateAdmin},
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("unable to fetch template admins: %w", err)
+	}
+
+	usersByIDs := map[uuid.UUID]database.GetUsersRow{}
+	var userIDs []uuid.UUID
+	for _, user := range users {
+		usersByIDs[user.ID] = user
+		userIDs = append(userIDs, user.ID)
+	}
+
+	var templateAdmins []database.GetUsersRow
+	if len(userIDs) > 0 {
+		orgIDsByMemberIDs, err := db.GetOrganizationIDsByMemberIDs(ctx, userIDs)
+		if err != nil {
+			return nil, xerrors.Errorf("unable to fetch organization IDs by member IDs: %w", err)
+		}
+
+		for _, entry := range orgIDsByMemberIDs {
+			if slices.Contains(entry.OrganizationIDs, stats.TemplateOrganizationID) {
+				templateAdmins = append(templateAdmins, usersByIDs[entry.UserID])
+			}
+		}
+	}
+	sort.Slice(templateAdmins, func(i, j int) bool {
+		return templateAdmins[i].Username < templateAdmins[j].Username
+	})
+
+	return templateAdmins, nil
+}
```

Comment:
1) What this function actually does is a bit hard to read. 
My read of it is:
- For a given template:
  - Query all users with the site-wide "template admin" role.
  - Fetch all organizations of which the above users are a member
  - Return only those users who are a member of that same organization as the given template.
Might make sense to add a clarifying comment?

2) It seems that this could be simplified significantly by adding the capability to query by `org_id` to `GetUsers`. I don't think this is a blocker, as my recollection is that modifying this query is rather non-trivial. It might be good to add as a follow-up though.

Reply:
AFAIR I _borrowed_ the logic from `templateACL` (and later `provisionedrserver.go`). I can definitely leave a comment describing what is going on here.

Re 2: I do not want to dive into org related stuff if it is not necessary :)